src/sciline/pipeline.py

```diff
@@ -452,11 +461,17 @@ def _get_provider(
             subproviders := self._subproviders.get(origin)
         ) is not None:
             requested = get_args(tp)
-            matches = [
-                (args, subprovider)
-                for args, subprovider in subproviders.items()
-                if _is_compatible_type_tuple(requested, args)
-            ]
+
+            match_groups = _groupby(
+                lambda m: sum(1 for t in m[0] if isinstance(t, TypeVar)),
+                (
+                    (args, subprovider)
+                    for args, subprovider in subproviders.items()
+                    if _is_compatible_type_tuple(requested, args)
+                ),
+            )
+            matches = match_groups[min(match_groups.keys(), default=0)]
```

Comment:
The use of defaultdict comes from the `groupby`, for all intents and purposes it could have been a regular dict.  In line 473 the providers with the fewest `TypeVar` arguments are passed on and the rest are dropped.

Reply:
> whole bunch of tests missing for generics with multiple typevars and various partial specializations.

Yeah you're right. I'll fix that.

Reply:
My point is: If there are no matches, then you would get a `KeyError` here, unless you use `defaultdict`, which is the wrong exception type, right?

Reply:
Yeah okay I see, the logic should be made clearer there, it's just a fluke that we don't get a bug because `match_groups` happens to be `defaultdict`.